external-crates/move/crates/move-compiler/src/command_line/compiler.rs

```diff
@@ -307,18 +310,29 @@ impl<'a> Compiler<'a> {
         {
             compilation_env.add_custom_known_filters(filters, filter_attr_name)?;
         }
-        let (source_text, pprog_and_comments_res) =
+
+        let (source_text, pprog, comments, mut diags) =
             parse_program(&mut compilation_env, maps, targets, deps)?;
-        let res: Result<_, Diagnostics> = pprog_and_comments_res.and_then(|(pprog, comments)| {
-            SteppedCompiler::new_at_parser(compilation_env, pre_compiled_lib, pprog)
-                .run::<TARGET>()
-                .map(|compiler| (comments, compiler))
-        });
-        Ok((source_text, res))
+
+        let env_result = compilation_env.check_diags_at_or_above_severity(Severity::BlockingError);
+        if no_fail || (diags.is_empty() && env_result.is_ok()) {
+            // diagnostics returned from parse_program are not in the compilation environment
```

Comment:
This is based on my intuition and the way the original implementation worked (where env diags were stuffed into diags returne from `parse_file`) and the existing tests pass. Still, it would be good to get a second opinion on whether this logic is correct.